README.md
nagios-plugins-openstack
============

## Description

Nagios probes to monitor an openstack cluster.

==============

Overview
--------

Deb package and PPA (example for 20140710 and catalystit/openstack)
-----------

```
cd ..
tar zcvf nagios-plugins-openstack_1.20140710.orig.tar.gz nagios-plugins-openstack
cd nagios-plugins-openstack
dpkg-buildpackage -rfakeroot -d -us -uc -S
cd ..
debsign nagios-plugins-openstack_1.20140710-1_source.changes
dput -f ppa:catalystit/openstack nagios-plugins-openstack_1.20140710-1_source.changes
```




debian/changelog
nagios-plugins-openstack (1.20140813.1-1) precise; urgency=low

  * Bug fixes:
     - #3: check_graphite.py not returning status properly

 -- Ricardo Rocha <ricardo@catalyst.net.nz>  Wed, 13 Aug 2014 12:34:37 +1300

nagios-plugins-openstack (1.20140716.2-1) precise; urgency=low

  * Added check_graphite.py

 -- Ricardo Rocha <ricardo@catalyst.net.nz>  Wed, 16 Jul 2014 13:34:37 +1300

nagios-plugins-openstack (1.20140710-1) precise; urgency=low

  * Added check_tempest.py probe

 -- Ricardo Rocha <ricardo@catalyst.net.nz>  Thu, 10 Jul 2014 11:01:37 +1300

nagios-plugins-openstack (1.20140416-1) unstable; urgency=low

  * Added event handler for automated nova evacuate

 -- Ricardo Rocha <ricardo@rocha-desk>  Wed, 16 Apr 2014 17:01:37 +1300

nagios-plugins-openstack (1.20140310-1) unstable; urgency=low

  * Removed token from check_keystone plugin output

 -- Ricardo Rocha <ricardo@rocha-desk>  Mon, 10 Mar 2014 17:01:37 +1300

nagios-plugins-openstack (1.20140220-1ubuntu1) saucy; urgency=low

  * Added cacert/insecure options to check_glance/keystone

 -- Ricardo Rocha <rocha@rocha-x11>  Thu, 20 Feb 2014 11:01:03 +1300

nagios-plugins-openstack (1.20120627-2) unstable; urgency=low

  * Add missing author from copyright file: Ignace Mouzannar
  
  * Use .install file instead of manually installing in the debian/rules file

 -- Loic Dachary (OuoU) <loic@debian.org>  Wed, 27 Jun 2012 15:27:19 +0200

nagios-plugins-openstack (1.20120627-1) unstable; urgency=low

  * Initial release (Closes: #675031).

 -- Loic Dachary (OuoU) <loic@debian.org>  Tue, 26 Jun 2012 23:25:06 +0200


plugins/check_glance.py
#!/usr/bin/env python
# -*- encoding: utf-8 -*-
#
# Keystone monitoring script for Nagios
#
# Copyright © 2012 eNovance <licensing@enovance.com>
#
# Author: Florian Lambert <florian.lambert@enovance.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#


import sys
import argparse

from keystoneclient.v2_0 import client as ksclient
import glanceclient as glance_client
from utils import EnvDefault

STATE_OK = 0
STATE_WARNING = 1
STATE_CRITICAL = 2
STATE_UNKNOWN = 3

def collect_args():

  parser = argparse.ArgumentParser(description='Check an OpenStack glance server.')
  parser.add_argument('--auth_url', metavar='URL', type=str, required=True,
        action=EnvDefault, envvar='OS_AUTH_URL', help='Keystone URL')
  parser.add_argument('--username', metavar='username', type=str, required=True,
        action=EnvDefault, envvar='OS_USERNAME', help='username to use for authentication')
  parser.add_argument('--password', metavar='password', type=str, required=True,
        action=EnvDefault, envvar='OS_PASSWORD', help='password to use for authentication')
  parser.add_argument('--tenant', metavar='tenant', type=str, required=True,
        action=EnvDefault, envvar='OS_TENANT_NAME', help='tenant name to use for authentication')
  parser.add_argument('--req_count', metavar='numberImages', type=str,
        required=False,
        help='minimum number of images in glance')
  parser.add_argument('--req_images', metavar='imagesName', type=str, nargs='+',
        required=False,
        help='name of images who must be available')
  parser.add_argument('--region_name', metavar='region_name', type=str,
        action=EnvDefault, envvar='OS_REGION_NAME', help='Region to select for authentication')
  parser.add_argument('--ca-cert', metavar='ca_cert', type=str,
                    action=EnvDefault, envvar='OS_CACERT', help='Location of the CA cert for validation')
  parser.add_argument('--insecure', action='store_true', default=False,
                    help='Do not verify certificates')
  return parser



def check_glance(c,args):
  #Flags resultat
  valid_image = 0
  count = 0


  if args.req_count:
    required_count = int(args.req_count)
    if len(list(c.images.list(**{"limit": required_count}))) >= required_count:
      count = 1

  #filters = {}
  #filters['name'] = "Debian GNU/Linux 6.0.4 amd64"
  #filters['container_format'] = "ami"

  if args.req_images:
    required_images = args.req_images

    for image in required_images:
      try:
        if len(list(c.images.list(**{"filters": {"name": image}}))) == 1:
          valid_image = valid_image + 1
      except :
        pass

  #parameters = {"filters": filters, "limit": limit}
  #images = c.get_images(**parameters)


  if args.req_count and count == 0:
    print "Failed - less than %d images found" % (required_count)
    sys.exit(STATE_CRITICAL)


  if args.req_images and valid_image < len(required_images):
    print "Failed - '%s' %d/%d images found " % (required_images, valid_image, len(required_images))
    sys.exit(STATE_WARNING)


  if args.req_images and args.req_count:
    print "OK - image %s found and enough images >=%d" % (required_images, required_count)
  elif args.req_images:
    print "OK - image %s found" % required_images
  elif args.req_count:
    print "OK - more than %d images found" % required_count
  else :
    print "OK - Connection glance established"


if __name__ == '__main__':
  args = collect_args().parse_args()
  try:
    ks_client = ksclient.Client(username=args.username,
                                password=args.password,
                                tenant_name=args.tenant,
                                auth_url=args.auth_url,
                                cacert=args.ca_cert,
                                insecure=args.insecure)

    token = ks_client.auth_token
    endpoint = ks_client.service_catalog.url_for(service_type='image')

    c = glance_client.Client('1', endpoint, token=token, cacert=args.ca_cert, insecure=args.insecure)
    sys.exit(check_glance(c, args))
  except Exception as e:
    print str(e)
    sys.exit(STATE_CRITICAL)


plugins/check_novaapi.py
#!/usr/bin/env python
# -*- encoding: utf-8 -*-
#
# Keystone monitoring script for Nagios
#
# Copyright © 2012 eNovance <licensing@enovance.com>
#
# Author: Florian Lambert <florian.lambert@enovance.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#


#needed python-novaclient
#https://github.com/openstack/python-novaclient

import sys
import argparse

from novaclient.v1_1 import client
from utils import EnvDefault

STATE_OK = 0
STATE_WARNING = 1
STATE_CRITICAL = 2
STATE_UNKNOWN = 3


#2 Warn = 1 Critical
def return_state(state):
  global RETURN_STATE
  global STATE_MESSAGE
  RETURN_STATE += state
  if RETURN_STATE > 1:
    STATE_MESSAGE +=" does not work"
    print STATE_MESSAGE
    sys.exit(STATE_CRITICAL)

def collect_args():
  parser = argparse.ArgumentParser(description='Check an OpenStack glance server.')
  parser.add_argument('--auth_url', metavar='URL', type=str, required=True,
        action=EnvDefault, envvar='OS_AUTH_URL', help='Keystone URL')
  parser.add_argument('--username', metavar='username', type=str, required=True,
        action=EnvDefault, envvar='OS_USERNAME', help='username to use for authentication')
  parser.add_argument('--password', metavar='password', type=str, required=True,
        action=EnvDefault, envvar='OS_PASSWORD', help='password to use for authentication')
  parser.add_argument('--tenant', metavar='tenant', type=str, required=True,
        action=EnvDefault, envvar='OS_TENANT_NAME', help='tenant name to use for authentication')
  return parser

def check_novaapi(nt):

  global RETURN_STATE
  global STATE_MESSAGE
  RETURN_STATE = STATE_OK
  STATE_MESSAGE = "Failed -"

  #flavors
  if not len(nt.flavors.list(detailed=False)) >= 1:
    STATE_MESSAGE +=" flavors.list >=1"
    return_state(STATE_WARNING)

  #servers
  if not nt.servers.list():
    STATE_MESSAGE +=" servers.list==false"
    return_state(STATE_WARNING)

  #images
  if not len(nt.images.list(detailed=False)) >= 1:
    STATE_MESSAGE +=" images.list >=1"
    return_state(STATE_WARNING)

  #security_groups
  if not len(nt.security_groups.list()) >= 1:
    STATE_MESSAGE +=" security_groups >=1"
    return_state(STATE_WARNING)



  if RETURN_STATE == STATE_WARNING:
    STATE_MESSAGE +=" does not work"
    print STATE_MESSAGE
  else:
    print "OK - Nova-api Connection established"

  return RETURN_STATE

if __name__ == '__main__':
  args = collect_args().parse_args()
  try:
    nt = client.Client(args.username,
         args.password,
         args.tenant,
         args.auth_url,
         service_type="compute")
    sys.exit(check_novaapi(nt))
  except Exception as e:
  	print str(e)
  	sys.exit(STATE_CRITICAL)


plugins/check_tempest.py
#!/usr/bin/env python
#
# vim: tabstop=2 shiftwidth=2
#
# Copyright (C) 2014 Catalyst IT Limited.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; only version 2 of the License is applicable.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
#
# Authors:
#   Ricardo Rocha <ricardo@catalyst.net.nz>
#
# About this plugin:
#   This nagios plugin launches one or more tempest tests, collects results
#   and sets the result state accordingly.
#
# Example usage:
#   ./check_tempest.py -l /usr/share/python-tempest -r 'test.api.image.v2.*'
#

import argparse
import nose
import sys
import traceback

STATE_OK = 0
STATE_WARNING = 1
STATE_CRITICAL = 2
STATE_UNKNOWN = 3

def collect_args():
  """
  Collects args passed in the cli.
  """
  parser = argparse.ArgumentParser(
          description='Executes tempest tests and collects results')
  parser.add_argument('-l', '--location', dest='location', type=str, action='store',
      required=True, help='Location of the tests to be run')
  parser.add_argument('-r', '--regexp', dest='regexp', type=str, action='store',
      required=True, help='Do not verify certificates')
  parser.add_argument('-w','--failiswarn', dest='failiswarn', action='store_true',
      help='return warn on failure (default is critical)')
  return parser

def check_tempest(args):
  success = nose.run(argv=[
    'nosetests', "-w%s" % args.location, "-m%s" % args.regexp,
  ])

  if success:
    return STATE_OK

  if args.failiswarn:
    return STATE_WARNING
  else:
    return STATE_CRITICAL

if __name__ == '__main__':
  args = collect_args().parse_args()
  try:
    check_tempest(args)
  except Exception as e:
    traceback.print_exc()
    sys.exit(STATE_CRITICAL)


eventhandlers/nova_evacuate_vms.py
#!/usr/bin/env python
# -*- encoding: utf-8 -*-
#
# Nagios event handler to evacuate VMs on a compute node failure.
#
# Copyright 2014 Catalyst IT.
#
# Author: Ricardo Rocha <ricardo@catalyst.net.nz>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import argparse
import random
import re
import sys
import syslog
from datetime import datetime, timedelta
import time
from keystoneclient.v2_0 import client as kclient
from keystoneclient import exceptions
from novaclient.v1_1 import client as nclient

syslog.openlog('nagios-nova-evacuate', 0, syslog.LOG_USER)

#
# State ID values for HOST states in nagios
#
UP = 0
DOWN = 1
UNREACHABLE = 2

#
# Argument parsing, as given by Nagios
#
parser = argparse.ArgumentParser(description='Evacuate VMs from compute node')
parser.add_argument('--auth_url', metavar='URL', type=str, required=True,
    help='Endpoint to the keystone service')
parser.add_argument('--username', metavar='username', type=str, required=True,
    help='username to use for authentication')
parser.add_argument('--password', metavar='password', type=str, required=True,
    help='password to use for authentication')
parser.add_argument('--tenant', metavar='tenant', type=str, required=True,
    help='tenant name to use for authentication')
parser.add_argument('--region_name', metavar='region_name', type=str,
    required=True, help='Region to select for authentication')
parser.add_argument('--ca-cert', metavar='ca_cert', type=str,
    help='Location of CA validation cert')
parser.add_argument('--insecure', action='store_true', default=False,
    help='Do not perform certificate validation')
parser.add_argument('--unreachable-is-down', action='store_true', default=False,
    help='True if we trigger evacuate on node unreachable')
parser.add_argument('--wait-timeout', metavar='wait_timeout', default=10,
    help='Time (in seconds) to wait for a successful evacuation before reporting failure')
parser.add_argument('compute_host', metavar='compute_host', type=str,
    help='Hostname of the compute node to evacuate')
parser.add_argument('state', metavar='state', type=str,
    help='Current state of probe (UP, DOWN, UNREACHEABLE)')
parser.add_argument('state_type', metavar='state_type', type=str,
    help='Current state type of probe (HARD, SOFT)')
args = parser.parse_args()

# By default unreachable does not trigger evacuate, but it's configurable
down_states = [DOWN]
if args.unreachable_is_down:
  down_states.append(UNREACHABLE)

# Is the state DOWN, and the state type HARD? Otherwise we do nothing for now...
if args.state_type != 'HARD' and args.state not in down_states:
  syslog.syslog("%s down, but probe not in HARD state yet (not running)" % args.compute_host)
  sys.exit(0)

syslog.syslog("%s DOWN and probe in HARD state, evacuating VMs" % args.compute_host)
# Get a nova client object (it takes care of keystone auth too)
try:
  nova = nclient.Client(args.username, args.password, args.tenant,
      auth_url=args.auth_url, region_name=args.region_name,
      insecure=args.insecure)
except Exception as e:
  syslog.syslog(syslog.LOG_ERR, "Failed to authenticate to keystone: %s" % str(e))
  sys.exit(-1)

# Check nova-compute is marked down for the compute host
for binary in ['nova-compute']:
  try:
    service_state = nova.services.list(host=args.compute_host, binary=binary)

    if len(service_state) != 1:
      syslog.syslog(syslog.LOG_ERR, "Got more than one %s on host %s" % (binary, args.compute_host))
      sys.exit(-1)
    if service_state.pop().state != 'down':
      syslog.syslog(syslog.LOG_ERR, "Nagios says down, but %s is still up in %s when querying nova" % (
          binary, args.compute_host))
      sys.exit(-1)
  except Exception as e:
    syslog.syslog(syslog.LOG_ERR, "Failed to list services %s for host %s :: %s" % (
      binary, args.compute_host, e))
    sys.exit(-1)

# List VMs associated to that compute node
try:
  vms = nova.servers.list(search_opts={'host': args.compute_host})
except Exception as e:
  syslog.syslog(syslog.LOG_ERR, "Failed to list VMs for host %s :: %s" % (args.compute_host, e))
  sys.exit(-1)

# Build target list (for the round robin approach)
# TODO: we should be using the scheduler instead of only a random assign
# https://blueprints.launchpad.net/nova/+spec/flexible-evacuate-scheduler
current_host = None
targets = []
try:
  services = nova.services.list()
  for service in services:
    if service.binary == 'nova-compute' and service.host != args.compute_host:
      targets.append(service.host)
except Exception as e:
  syslog.syslog(syslog.LOG_ERR, "Failed while filtering target hosts from service list :: %s" % e)
  sys.exit(-1)

# Trigger evacuation for each VM
# We collect the results to build a final report
results = {'success': [], 'failures': []}
for vm in vms:
  target = random.choice(targets)
  syslog.syslog("Evacuating '%s' to compute host '%s'" % (vm.name, target))
  try:
    success = False
    vm.evacuate(target, True)
    # wait for ACTIVE, or give up after 30 seconds
    start = datetime.now()
    cvm = nova.servers.list(search_opts={'name': vm.name})
    while datetime.now() < (start + timedelta(seconds=args.wait_timeout)):
      if cvm[0].status == 'ACTIVE':
        results['success'].append((vm.name, target))
        break
      time.sleep(3)
      cvm = nova.servers.list(search_opts={'name': vm.name})
    if cvm[0].status != 'ACTIVE':
      results['failures'].append((vm.name, "VM is in ERROR or UNKNOWN status, needs manual migration"))
  except Exception as e:
    results['failures'].append((vm.name, str(e)))
    syslog.syslog(syslog.LOG_ERR, "Failed to evacuate vm '%s' :: %s" % (vm.name, e))
syslog.syslog(syslog.LOG_ERR, "Evacuation of %s :: Successes (%d, %s) :: Failures (%d, %s)" % (
  args.compute_host, len(results['success']), results['success'], len(results['failures']), results['failures']))


plugins/check_graphite.py
#!/usr/bin/env python
#
# vim: tabstop=2 shiftwidth=2
#
# Copyright (C) 2014 Catalyst IT Limited.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; only version 2 of the License is applicable.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
#
# Authors:
#   Ricardo Rocha <ricardo@catalyst.net.nz>
#
# About this plugin:
#   This nagios plugin executes a graphite query, checking that results are
#   retrieved. Useful to make sure expected data is being published.
#
#   Check: http://graphite.readthedocs.org/en/1.0/url-api.html
#   for more information on the 'target' and 'from' options.
#
# Example usage:
#   ./check_graphite.py -H icinga.example.com -P 80 -p http -t 'collectd.*.ceph-ceph.cluster.gauge.total_avail' -f '-10minutes'
#

import argparse
import json
import sys
import traceback
import urllib2

STATE_OK = 0
STATE_WARNING = 1
STATE_CRITICAL = 2
STATE_UNKNOWN = 3

def collect_args():
  """
  Collects args passed in the cli.
  """
  parser = argparse.ArgumentParser(
    description='Executes a graphite query and checks for results')
  parser.add_argument('-H', '--host', dest='host', type=str, action='store',
    required=True, help='Graphite host to contact')
  parser.add_argument('-P', '--port', dest='port', type=str, action='store',
    default='80', help='Port the graphite daemon listens')
  parser.add_argument('-p', '--proto', dest='proto', type=str, action='store',
    default='http', help='Protocol to use (one of http/https)')
  parser.add_argument('-s', '--subpath', dest='subpath', type=str, action='store',
    default='', help='Subpath of graphite the host url')
  parser.add_argument('-t', '--target', dest='target', type=str, action='store',
    required=True, help='Target query (the metric to query for)')
  parser.add_argument('-f', '--from', dest='interval', type=str, action='store',
    required=True, help='Interval to query for (check \'from\' in graphite)')
  parser.add_argument('-w','--failiswarn', dest='failiswarn', action='store_true',
      help='Return warn on failure (default is critical)')
  parser.add_argument('-v','--verbose', dest='verbose', action='store_true',
      help='Print some additional information')
  return parser

def check_graphite(args):

  query = "%s://%s:%s/%srender/?target=%s&from=%s&format=json" % (
    args.proto, args.host, args.port, args.subpath, args.target, args.interval)

  result = urllib2.urlopen(query).read()
  if args.verbose:
    print result
  jsonres = json.loads(result)

  if len(jsonres) > 0 and len(jsonres[0]['datapoints']) > 0:
    print "OK: query %s retrieved %d elements" % (query, len(jsonres[0]['datapoints']))
    return STATE_OK

  print "Failed: query %s returned %s" % (query, result)

  if args.failiswarn:
    return STATE_WARNING
  else:
    return STATE_CRITICAL

if __name__ == '__main__':
  args = collect_args().parse_args()
  try:
    sys.exit(check_graphite(args))
  except Exception as e:
    print "Failed: %s" % str(e)
    sys.exit(STATE_CRITICAL)
